### PR TITLE
[WIP] full finetune / qlora + ac/offload/optm in bwd

### PR DESCRIPTION
**Pull request by weifengpy**
## Why composing FSDP with NF4Tensor

**QLoRA** : number of trainable parameters are reduced from xxx to xxx. parameter size are reduced by xx
Full finetuning original Llama with 4bit quantized params:

7B + QLora on FFNs: summarizing memory usage below with bf16, adamW, AC, cpu offloading
* sharding NF4Tensor in FSDP: NF4Tensor are 4 bit quant weight from QLora
* cpu offloading NF4Tensor in FSDP: most profitabble
* gradient in the backward, 8bit optimizer: does not matter in QLora because of tiny gradable parameters. should priooritize in full training

Screenshot 2024-02-26 at 12 50 45 PM

Screenshot 2024-02-27 at 12 39 54 PM

### CODE REVIEW

**Comment by awgu** [comment]
Should we call out that this table assumes that we are only applying QLoRA to the FFNs?

File: transformer_nuggets/quant/nf4_tensor.py [hunk 1]
```diff
@@ -1,13 +1,303 @@
...
+
+@implements(
+    [
+        aten.detach.default,
+    ]
+)
+def nf4_detach(aten_op, args, kwargs=None):
+    # nn.Parameter need detach
+    quantized_data = aten_op(args[0].quantized_data, *args[1:], **kwargs)
+    tensor_meta = SubclassTensorArgs(
```
Comment (drisspg) [inline transformer_nuggets/quant/nf4_tensor.py:32]: https://github.com/drisspg/transformer_nuggets/pull/24/files#diff-6954806b06a395823616c5a0fefc77b5cf274b3f88a5381c38856cbf74d6f2bfR37-R45

I hijacked the tensor unflatten/flatten to construct the new tensors,

**Comment by weifengpy** [comment]
close is since QLoRA + FSDP2 has been landed into torchtune: https://github.com/pytorch/torchtune/pull/909
